File: 917.go
import "unicode"

func reverseOnlyLetters(s string) string {
	start, end := 0, len(s)-1
	result := []byte(s)
	for {
		for !unicode.IsLetter(rune(result[start])) && start < end {
			start++
		}
		for !unicode.IsLetter(rune(result[end])) && start < end {
			end--
		}
		if start >= end {
			return string(result)
		}

		result[start], result[end] = result[end], result[start]
		start++
		end--
	}
}

File: 917.rs
impl Solution {
    pub fn reverse_only_letters(mut s: String) -> String {
        unsafe {
            let mut s = s.as_bytes_mut();
    
            let mut start = 0;
            let mut end = s.len() - 1;
            while start < end {
                if s[start] >= b'A' && s[start] <= b'Z' || s[start] >= b'a' && s[start] <= b'z' {
                    while !(s[end] >= b'A' && s[end] <= b'Z' || s[end] >= b'a' && s[end] <= b'z') {
                        end -= 1;
                    }
                    s.swap(start, end);
                    end -= 1;
                }
                start += 1;
            }
        }
        s
    }
}
